connection_go18.go

```diff
@@ -0,0 +1,234 @@
+		return nil, err
+	}
+
+	var err error
+	var tx driver.Tx
+	if opts.ReadOnly {
+		tx, err = mc.beginReadOnly()
+	} else {
+		tx, err = mc.Begin()
+	}
+	mc.finish()
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	default:
+	case <-ctx.Done():
+		tx.Rollback()
+		return nil, ctx.Err()
+	}
+	return tx, err
+}
+
+func (mc *mysqlConn) beginReadOnly() (driver.Tx, error) {
+	if mc.isBroken() {
+		errLog.Print(ErrInvalidConn)
+		return nil, driver.ErrBadConn
+	}
+	// https://dev.mysql.com/doc/refman/5.7/en/innodb-performance-ro-txn.html
+	err := mc.exec("START TRANSACTION READ ONLY")
+	if err != nil {
+		return nil, err
+	}
+
+	return &mysqlTx{mc}, nil
+}
+
+func (mc *mysqlConn) QueryContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Rows, error) {
+	dargs, err := namedValueToValue(args)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := mc.watchCancel(ctx); err != nil {
+		return nil, err
+	}
+
+	rows, err := mc.Query(query, dargs)
+	if err != nil {
+		mc.finish()
+		return nil, err
+	}
+	if set, ok := rows.(setfinish); ok {
+		set.setFinish(mc.finish)
+	} else {
+		mc.finish()
+	}
+	return rows, err
+}
+
+func (mc *mysqlConn) ExecContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Result, error) {
+	dargs, err := namedValueToValue(args)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := mc.watchCancel(ctx); err != nil {
+		return nil, err
+	}
+	defer mc.finish()
+
+	return mc.Exec(query, dargs)
+}
+
+func (mc *mysqlConn) PrepareContext(ctx context.Context, query string) (driver.Stmt, error) {
+	if err := mc.watchCancel(ctx); err != nil {
+		return nil, err
+	}
+
+	stmt, err := mc.Prepare(query)
+	mc.finish()
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	default:
+	case <-ctx.Done():
+		stmt.Close()
+		return nil, ctx.Err()
+	}
+	return stmt, nil
+}
+
+func (stmt *mysqlStmt) QueryContext(ctx context.Context, args []driver.NamedValue) (driver.Rows, error) {
+	dargs, err := namedValueToValue(args)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := stmt.mc.watchCancel(ctx); err != nil {
+		return nil, err
+	}
+
+	rows, err := stmt.Query(dargs)
+	if err != nil {
+		stmt.mc.finish()
+		return nil, err
+	}
+	if set, ok := rows.(setfinish); ok {
+		set.setFinish(stmt.mc.finish)
+	} else {
+		stmt.mc.finish()
+	}
+	return rows, err
+}
+
+func (stmt *mysqlStmt) ExecContext(ctx context.Context, args []driver.NamedValue) (driver.Result, error) {
+	dargs, err := namedValueToValue(args)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := stmt.mc.watchCancel(ctx); err != nil {
+		return nil, err
+	}
+	defer stmt.mc.finish()
+
+	return stmt.Exec(dargs)
+}
+
+func (mc *mysqlConn) watchCancel(ctx context.Context) error {
+	select {
+	default:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	if mc.watcher == nil {
+		return nil
+	}
+
+	mc.watcher <- ctx
+
+	return nil
+}
+
+func (mc *mysqlConn) startWatcher() {
+	watcher := make(chan mysqlContext, 1)
```

Comment:
I also have some reservations about all connections sharing this same watcher goroutine.

On one hand, it's good that there's a single goroutine for doing this and we can avoid the overhead of goroutine creation every time we're running a query or whatever.

On the other, it means that only one connection can be watched at a time. And because the `watcher` channel has a buffer size of 1, only two connections can be running queries at the same time. You could increase the buffer size of `watcher`, but even if you do that only one connection will be watched for cancellation at a time.

For example, if you have an intentionally long-running query with a long timeout that's on the watched connection and you also have an **unintentionally** long-running query that sits in the `watcher` buffer and will be canceled, the one sitting in the buffer has to wait until the first query finishes before it can realize that it's run too long and should be canceled.

I'm not sure what concrete suggestions to offer here, but I think it's worth mentioning.